### ISSUE TITLE
Adding example for transformation between coordinate systems

### ISSUE THREAD

**Issue by Sebastian-Schuetz** [moble/quaternion#165]
Hi there,
first of all thank you for this amazing package and the clear warning concerning euler angles. 
Without it, I might have already thrown the quaternions over board and tryed my luck with those,
but they seem to be a pact with the devil.

I need to accomplish a transform of rotation of a point D in coordinate system A to its corresponding representation in coordinate system B with two seperate origins. 
A and B share a common point C (but not the origin)
I know the representation of C in A and B.

So far, I thought about the following approach:

- get the rotation between C and D in coordinate System A
- get the rotation between the two coordinate Systems in point C
- transform the rotation in coordinate system B, by applying the rotation between the two coordinate systems
- apply the transformed rotation to point C in B

```python
import numpy as np
import quaternion

C_rot_in_A = quaternion.quaternion(0.67548805475235, -0.214063242077827, 0.225458413362503, -0.668626487255096)
D_rot_in_A = quaternion.quaternion(0.629053235054016, -0.182543635368347, 0.289672702550888, -0.697896599769592)
C_rot_in_B = quaternion.quaternion(0.00632629987645627, -0.585520499936184, 0.808567045571855, 0.0578364495759909)

# wanted, but not known...
D_rot_in_B = quaternion.quaternion(0.0898282907587203, -0.578358930098212, 0.806664413554814, 0.0820021344800612)

# get rotation between point C and D in A
rotation_C_to_D_in_A = C_rot_in_A * D_rot_in_A.inverse()

# get rotation between the coordinate systems
rotation_coordinate_system = C_rot_in_A * C_rot_in_B.inverse()

# transform the rotation between C and D into coordinate system B
rotation_C_to_D_in_B = rotation_C_to_D_in_A * rotation_coordinate_system
# apply the transformed rotation to the original rotation in point C in B
rotation_D_in_B = C_rot_in_B * rotation_C_to_D_in_B

print(rotation_D_in_B)
# outputs: quaternion(0.716431228247433, -0.037980269746783, 0.215762589206211, 0.662367215490306)
# expected: quaternion(0.0898282907587203, -0.578358930098212, 0.806664413554814, 0.0820021344800612)

```

**Describe the solution you'd like**
But as you can see, this results in the wrong output. I can´t put my finger on the why (sadly I am lacking the mathematical background...)

But as it seems, the request to rotate objects seems to be quite often: e.g. #163 #151 #149 #29
Maybe one could add some examples, for "common" rotations/operations?

**Describe alternatives you've considered**
As an alternative one could link the issues in the wiki, sorted by some headings?
Personally I think they are a very helpful resource and the explanation at #151 really helped me a lot.

**Comment by moble** [comment]
Unfortunately, it doesn't look like you have (or have not told me about) enough information to actually solve this problem.  In particular, just knowing the coordinates of one point `C` in coordinate systems `A` and `B` is not sufficient to relate those two coordinate systems to each other.  I could rotate `A` about the line joining its origin and `C`, and the coordinates of `C` would not change; ditto rotating `B`.  So just the coordinates of `C` in both systems are not enough.  You would need some other criterion and/or some other points measured in both systems.

---

It's also not clear to me how you're trying to use quaternions.  What do all these `[X]_rot_in_[Y]` variables represent?  You said that `C` was a point, but `C_rot_in_A` is a (nearly) unit quaternion.  If `C` is a point, its most natural representation would be as a vector.

And you say that `A` and `B` have different origins.  Basically, this means that you'll have to *subtract* vectors to account for the displacement between their origins.  I don't see where you're doing that.

Is it actually possible to *rotate* "between C and D in coordinate System A"?  This would mean that those two points are the same distance from the origin of `A`.  Is it possible to *rotate* "between the two coordinate Systems in point C"?  This would mean that the origin of `A` and the origin of `B` are the same distance from `C`.

Basically, none of this is possible to do using *just* rotations except in very special situations that you haven't specified.

**Comment by Sebastian-Schuetz** [comment]
Thirst of, thanks for the answer. I am sorry, that my explanation wasn´t clear enough. I will try again to restate my problem.
I know the position and orientation of point C in the two different coordinate systems A and B and would need to transfer the position and orientation of point D from coordinate system A to coordinate system B.
I already was able to transfere the position of D into B, but couldn´t wrap my head around the rotations. So I excluded the
position-stuff, hoping to make a shorter "working example" (obviously didn´t work out)

The [X]_rot_in_[Y] represent the rotation/orientation of the point [X] in its coresponding coordinate system [Y].
Same goes for [X]_pos_in_[Y], just for the positions.

>Is it actually possible to rotate "between C and D in coordinate System A"? This would mean that those two points are the same distance from the origin of A.  Is it possible to rotate "between the two coordinate Systems in point C"? This would mean that the origin of A and the origin of B are the same distance from C

My rational was, that I could handle this similar to the position handling: Instead of a vector between the two points, I would use the rotation between the two points.The two points are not the same distance from the origin of A, nether are the origin of A and B the same distance from C. Would I need to use a affine transformation?

``
import numpy as np
import quaternion

"""
Definition of points
"""

C_pos_in_A = np.array([0.09207182377576828, 
                       -0.149169921875, 
                       0.08073931187391281])
C_rot_in_A = quaternion.quaternion(0.67548805475235, -0.214063242077827, 0.225458413362503, -0.668626487255096)

D_pos_in_A = np.array([0.06779178977012634,
                       -0.10643985122442245,
                       0.10021161288022995])
D_rot_in_A = quaternion.quaternion(0.629053235054016, -0.182543635368347, 0.289672702550888, -0.697896599769592)

C_pos_in_B = np.array([0.28065911694532897,
                       -0.3563198956163062,
                       0.15482871437395135])
C_rot_in_B = quaternion.quaternion(0.00632629987645627, -0.585520499936184, 0.808567045571855, 0.0578364495759909)

# wanted, but not known...
D_pos_in_B = np.array([0.3090784810880993,
                       -0.32058295002432974,
                       0.16085691024452017])
D_rot_in_B = quaternion.quaternion(0.0898282907587203, -0.578358930098212, 0.806664413554814, 0.0820021344800612)

"""
Logic for transforming the position
"""

coordinate_rot = C_rot_in_A.normalized() * C_rot_in_B.normalized()
coordinate_rot = coordinate_rot.normalized()

# get the translation vector in the A coordinate system
translation_in_A = np.subtract(D_pos_in_A, C_pos_in_A)

# rotate the vector into the B coordinate-system
translation_in_B = coordinate_rot.conjugate() * quaternion.quaternion(0, translation_in_A[0], translation_in_A[1], translation_in_A[2]) * coordinate_rot
translation_in_B = translation_in_B.vec

# Applying some correction to the vector...
# Mainly flipping some axis...
p = translation_in_B[0]
translation_in_B[0] = translation_in_B[1]
translation_in_B[1] = p
translation_in_B[2] *= -1

# Add the vector to point A
D_in_B = np.add(C_pos_in_B, translation_in_B)
print(D_in_B)
# outputs: [ 0.31227246 -0.3139747   0.15625377]
# expected: [0.3090784810880993, -0.32058295002432974, 0.16085691024452017])
# deviation properly due to misaligment in the assemby

"""
Logic for transforming the orientation
- not working -
"""

# get rotation between point C and D in A
rotation_C_to_D_in_A = C_rot_in_A * D_rot_in_A.inverse()

# get rotation between the coordinate systems
rotation_coordinate_system = C_rot_in_A * C_rot_in_B.inverse()

# transform the rotation between C and D into coordinate system B
rotation_C_to_D_in_B = rotation_C_to_D_in_A * rotation_coordinate_system
# apply the transformed rotation to the original rotation in point C in B
rotation_D_in_B = C_rot_in_B * rotation_C_to_D_in_B

print(rotation_D_in_B)
# outputs: quaternion(0.716431228247433, -0.037980269746783, 0.215762589206211, 0.662367215490306)
# expected: quaternion(0.0898282907587203, -0.578358930098212, 0.806664413554814, 0.0820021344800612)
``

**Comment by moble** [comment]
> I know the position and orientation of point C

Okay, this is why it's confusing.  A "point" is defined to have no spatial extent or orientation, so that's the wrong word.  It would be better to just think of C and D as coordinate systems, complete with basis vectors of their own.  Go back to ignoring the translations; you can pretend for now that the origins of A, B, C, and D are all the same.

So the good news is that now it sounds like you do have enough information to solve this problem.  But I'm still not sure what you mean by any of this.

> The `[X]_rot_in_[Y]` represent the rotation/orientation of the point `[X]` in its corresponding coordinate system `[Y]`.

I can imagine many ways to interpret that statement, even if I correct "point" to something more meaningful.

Are all the numbers that you used measured in a single coordinate system?  Note that, just as vectors have different components in different coordinate systems, so do quaternions.  (I'll just point out that these problems you're running into would also pop up with Euler angles, but be harder to diagnose.)

One problem is that your `C_rot_in_A` and `D_rot_in_A` are fairly close in value, and you seem to expect pretty large errors, so I can't tell how significant the difference between them is.  It would be better to start with nice, round, and very different values:
```python
C_rot_in_A = quaternion.quaternion(1, 0, 0, 1).normalized()  # rotation by pi/2 about z axis
D_rot_in_A = quaternion.quaternion(1, 0, 1, 0).normalized()  # rotation by pi/2 about y axis
C_rot_in_B = quaternion.quaternion(1, 1, 0, 0).normalized()  # rotation by pi/2 about x axis
```
Another problem is that I don't understand why you expect `D_rot_in_B` to have the value that you put in your code.  What would you expect it to be if you start with the nice values above, and why?

**Comment by Sebastian-Schuetz** [comment]
Ah, ok. So would pose (consisting of position and orientation) be a better description?

> But I'm still not sure what you mean by any of this.

I meant the representation of the pose, relative to the origin of the corresponding coordinate system. 
Meaning each pose is translated by [x]_pos_in_[y] from the origin and oriented [x]_ort_in_[y] relative to the orientation of the origin in its corresponding coordinate system. I hope this makes it more clear.

> Are all the numbers that you wrote in your code measured in a single coordinate system? 
No, they are messured in two different coordinate systems. The variable names ending in A being messured in coordinate system A, same with B.

On to the values: the values come from a 3D camera system and a robot arm, so I can´t change all of them.
Currently I get the expected values by moving the robot arm and reading out the pose in both coordinate systems.
(So I set C and D in coordinate system B and observed the pose with the camera)
The goal however is, that I can give the pose in the camera coordinate system and have the robot reach this pose.
(With C being the current position of the arm and D being the goal to reach)

```
import numpy as np
import quaternion

"""
Definition of points
"""

C_pos_in_A = np.array([-0.13001029193401337,
                       -0.3125380277633667,
                       0.14449119567871094])
C_rot_in_A = quaternion.quaternion(0.954804480075836, -0.0321213528513908, 0.249238103628159, -0.158735781908035)

D_pos_in_A = np.array([-0.0515381321310997,
                       -0.27589818835258484,
                       0.15720322728157043])
D_rot_in_A = quaternion.quaternion(-0.158271878957748, -0.255194544792175, -0.052324652671814, -0.952411651611328)

C_pos_in_B = np.array([0.20000054725812696,
                       -0.39999997376971413,
                       0.14999943583689282])
C_rot_in_B = quaternion.quaternion(9.7675712181119e-07, 4.24979496024042e-13, 0.999999999998729, 1.25957570638853e-06)


# wanted, but not known...
D_pos_in_B = np.array([0.2500000175368012,
                       -0.44999998990803713,
                       0.19999908593788587])
D_rot_in_B = quaternion.quaternion(5.97680036138157e-07, -0.999999999997815, 1.22172947138444e-06, -1.58735114010876e-06)

"""
Logic for transforming the position
"""

coordinate_rot = C_rot_in_A.normalized() * C_rot_in_B.normalized()
coordinate_rot = coordinate_rot.normalized()

# get the translation vector in the A coordinate system
translation_in_A = np.subtract(D_pos_in_A, C_pos_in_A)

# rotate the vector into the B coordinate-system
translation_in_B = coordinate_rot.conjugate() * quaternion.quaternion(0, translation_in_A[0], translation_in_A[1], translation_in_A[2]) * coordinate_rot
translation_in_B = translation_in_B.vec

# Applying some correction to the vector...
# Mainly flipping some axis...
p = translation_in_B[0]
translation_in_B[0] = translation_in_B[1]
translation_in_B[1] = p
translation_in_B[2] *= -1

# Add the vector to point A
D_in_B = np.add(C_pos_in_B, translation_in_B)
print(D_in_B)
# outputs: [ 0.25546313 -0.44715469  0.19860281]
# expected: [0.2500000175368012, -0.44999998990803713, 0.19999908593788587]
# deviation properly due to misaligment in the assemby

"""
Logic for transforming the orientation
- not working -
"""

# get rotation between point C and D in A
rotation_C_to_D_in_A = C_rot_in_A * D_rot_in_A.inverse()

# get rotation between the coordinate systems
rotation_coordinate_system = C_rot_in_A * C_rot_in_B.inverse()

# transform the rotation between C and D into coordinate system B
rotation_C_to_D_in_B = rotation_C_to_D_in_A * rotation_coordinate_system
# apply the transformed rotation to the original rotation in point C in B
rotation_D_in_B = C_rot_in_B * rotation_C_to_D_in_B

print(rotation_D_in_B)
# outputs: quaternion(0.149142048949535, -0.255499822821174, 0.0499438436362146, -0.953929845415057)
# expected: quaternion(5.97680036138157e-07, -0.999999999997815, 1.22172947138444e-06, -1.58735114010876e-06)
```

**Comment by moble** [comment]
> oriented [x]ort_in[y] relative to the orientation of the origin

Again, the origin is just a point, so it can't have an orientation.  And that still doesn't mean a whole lot.  I'm guessing that `[X]_rot_in_[Y]` represents the components of a quaternion in the `[Y]` coordinate system that rotates any of the `[Y]` basis vectors onto the corresponding `[X]` basis vector.  So, again, I'll emphasize that there are two possible differences between `C_rot_in_A` and `C_rot_in_B`.  First, the describe how to get to `C` from two different starting points.  Second, they are possibly expressed in two different coordinate systems.  That is, *quaternions themselves mean different things when expressed in `A` and in `B`.*  But you can rotate them just as if they were vectors.  If `R` is a quaternion that rotates the `A` coordinates onto the `B` coordinates, then you really want to compare`C_rot_in_A` to `R*C_rot_in_B*R.conjugate()` (or maybe the other way around, depending on how you interpret things).

**Comment by Sebastian-Schuetz** [comment]
Now I am starting to get my head wrapped around this. 
That should to the trick.
Thank you for your time and patience with me!
